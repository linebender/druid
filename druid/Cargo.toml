[package]
name = "druid"
version = "0.7.0"
license = "Apache-2.0"
authors = ["Raph Levien <raph.levien@gmail.com>"]
description = "Data-oriented Rust UI design toolkit."
repository = "https://github.com/linebender/druid"
categories = ["gui"]
readme = "README.md"
keywords = ["gui", "ui", "toolkit"]
edition = "2018"

[package.metadata.docs.rs]
# The "svg" and "image" features have doc clashes that cause undefined output in docs.
# See https://github.com/rust-lang/cargo/issues/6313 for more information.
# Once cargo doc becomes smart enough to handle multiple versions of the same crate,
# the "svg" and "image" features should be enabled for the docs.rs output.
features = ["im"]
rustdoc-args = ["--cfg", "docsrs"]
default-target = "x86_64-pc-windows-msvc"

[features]
default = ["gtk"]
gtk = ["druid-shell/gtk"]
image = ["druid-shell/image"]
svg = ["usvg", "harfbuzz-sys"]
x11 = ["druid-shell/x11"]
crochet = []
serde = ["im/serde"]

# passing on all the image features. AVIF is not supported because it does not
# support decoding, and that's all we use `Image` for.
png = ["druid-shell/image_png"]
jpeg = ["druid-shell/jpeg"]
jpeg_rayon = ["druid-shell/jpeg_rayon"]
gif = ["druid-shell/gif"]
bmp = ["druid-shell/bmp"]
ico = ["druid-shell/ico"]
tiff = ["druid-shell/tiff"]
webp = ["druid-shell/webp"]
pnm = ["druid-shell/pnm"]
dds = ["druid-shell/dds"]
tga = ["druid-shell/tga"]
farbfeld = ["druid-shell/farbfeld"]
dxt = ["druid-shell/dxt"]
hdr = ["druid-shell/hdr"]

# Remember to update this when changing an image feature.
image-all = ["image", "svg", "png", "jpeg", "jpeg_rayon", "gif", "bmp", "ico", "tiff", "webp", "pnm", "dds", "tga", "farbfeld", "dxt", "hdr"]

[dependencies]
druid-shell = { version = "0.7.0", default-features = false, path = "../druid-shell" }
druid-derive = { version = "0.4.0", path = "../druid-derive" }

tracing = { version = "0.1.22" }
tracing-subscriber = { version = "0.2.15", features = ["fmt", "ansi"], default-features = false }
fluent-bundle = "0.12.0"
fluent-langneg = "0.13.0"
fluent-syntax = "0.9.3"
unic-langid = "0.9.0"
unicode-segmentation = "1.6.0"
xi-unicode = "0.3.0"
fnv = "1.0.7"
instant = { version = "0.1.6", features = ["wasm-bindgen"] }

# Optional dependencies
im = { version = "15.0.0", optional = true }
usvg = { version = "0.12.0", optional = true }
# We don't depend directly on harfbuzz-sys, it comes through usvg. But we need to pin
# the version, because the latest release of harfbuzz_rs is broken with harfbuzz-sys 0.5.
harfbuzz-sys = { version = "0.4.0", optional = true }

[target.'cfg(target_arch="wasm32")'.dependencies]
tracing-wasm = { version = "0.1.0" }
console_error_panic_hook = { version = "0.1.6" }

[dev-dependencies]
float-cmp = { version = "0.8.0", features = ["std"], default-features = false }
# tempfile 3.2.0 broke wasm; I assume it will be yanked (Jan 12, 2021)
tempfile = "=3.1.0"
piet-common = { version = "=0.3.2", features = ["png"] }
pulldown-cmark = { version = "0.8", default-features = false }
test-env-log = { version = "0.2.5", features = ["trace"], default-features = false }

[[example]]
name = "cursor"
required-features = ["image", "png"]

[[example]]
name = "image"
required-features = ["image", "png"]

[[example]]
name = "invalidation"
required-features = ["im"]

[[example]]
name = "list"
required-features = ["im"]

[[example]]
name = "svg"
required-features = ["svg"]

[[example]]
name = "tabs"
required-features = ["im"]

[[example]]
name = "widget_gallery"
required-features = ["svg", "im", "image", "png"]
