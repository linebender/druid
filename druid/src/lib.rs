// Copyright 2018 The Druid Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Simple data-oriented GUI.
//!
//! Druid lets you build simple interactive graphical applications that
//! can be deployed on Windows, macOS, Linux, and the web.
//!
//! Druid is built on top of [`druid-shell`], which implements all of the
//! lower-level, platform-specific code, providing a common abstraction
//! for things like key and mouse events, creating windows, and launching
//! an application. Below [`druid-shell`] is [`piet`], which is a cross-platform
//! 2D graphics library, providing a simple and familiar drawing API that can be
//! implemented for various platforms.
//!
//! Druid is a data-driven, declarative framework. You describe your application
//! model in terms of the [`Data`] trait, and then you build up a tree of
//! [`widget`] s that can display and modify your data.
//!
//! Your widgets handle [`Event`]s, such as mouse movement, and can modify the data;
//! these changes are then delivered to relevant widgets, which can update
//! their state and redraw.
//!
//! As your application grows, you can use [`Lens`]es to expose only certain
//! subsets of your data model to certains subsets of your widget tree.
//!
//! For more information you should read the [druid book].
//!
//! # Examples
//!
//! For many more examples, see [`druid/examples`].
//!
//! ```no_run
//! use druid::widget::{Align, Flex, Label, TextBox};
//! use druid::{AppLauncher, Data, Env, Lens, LocalizedString, Widget, WindowDesc, WidgetExt};
//!
//! const VERTICAL_WIDGET_SPACING: f64 = 20.0;
//! const TEXT_BOX_WIDTH: f64 = 200.0;
//! const WINDOW_TITLE: LocalizedString<HelloState> = LocalizedString::new("Hello World!");
//!
//! #[derive(Clone, Data, Lens)]
//! struct HelloState {
//!     name: String,
//! }
//!
//! fn main() {
//!     // describe the main window
//!     let main_window = WindowDesc::new(build_root_widget())
//!         .title(WINDOW_TITLE)
//!         .window_size((400.0, 400.0));
//!
//!     // create the initial app state
//!     let initial_state = HelloState {
//!         name: "World".into(),
//!     };
//!
//!     // start the application
//!     AppLauncher::with_window(main_window)
//!         .launch(initial_state)
//!         .expect("Failed to launch application");
//! }
//!
//! fn build_root_widget() -> impl Widget<HelloState> {
//!     // a label that will determine its text based on the current app data.
//!     let label = Label::new(|data: &HelloState, _env: &Env| format!("Hello {}!", data.name));
//!     // a textbox that modifies `name`.
//!     let textbox = TextBox::new()
//!         .with_placeholder("Who are we greeting?")
//!         .fix_width(TEXT_BOX_WIDTH)
//!         .lens(HelloState::name);
//!
//!     // arrange the two widgets vertically, with some padding
//!     let layout = Flex::column()
//!         .with_child(label)
//!         .with_spacer(VERTICAL_WIDGET_SPACING)
//!         .with_child(textbox);
//!
//!     // center the two widgets in the available space
//!     Align::centered(layout)
//! }
//! ```
//!
//! # Optional Features
//!
//! * `im` - Efficient immutable data structures using the [`im` crate],
//!          which is made available via the [`im` module].
//! * `svg` - Scalable Vector Graphics for icons and other scalable images using the [`usvg` crate].
//! * `image` - Bitmap image support using the [`image` crate].
//! * `x11` - Work-in-progress X11 Linux backend instead of GTK.
//!
//! Features can be added with `cargo`. For example, in your `Cargo.toml`:
//! ```no_compile
//! [dependencies.druid]
//! version = "0.7.0"
//! features = ["im", "svg", "image"]
//! ```
//!
//! [`Widget`]: trait.Widget.html
//! [`Data`]: trait.Data.html
//! [`Lens`]: trait.Lens.html
//! [`widget`]: ./widget/index.html
//! [`Event`]: enum.Event.html
//! [`druid-shell`]: https://docs.rs/druid-shell
//! [`piet`]: https://docs.rs/piet
//! [`druid/examples`]: https://github.com/linebender/druid/tree/v0.7.0/druid/examples
//! [druid book]: https://linebender.org/druid/
//! [`im` crate]: https://crates.io/crates/im
//! [`im` module]: im/index.html
//! [`usvg` crate]: https://crates.io/crates/usvg
//! [`image` crate]: https://crates.io/crates/image

#![deny(
    broken_intra_doc_links,
    unsafe_code,
    clippy::trivially_copy_pass_by_ref
)]
#![warn(missing_docs)]
#![allow(clippy::new_ret_no_self, clippy::needless_doctest_main)]
#![cfg_attr(docsrs, feature(doc_cfg))]

// Allows to use macros from druid_derive in this crate
extern crate self as druid;
pub use druid_derive::Lens;

use druid_shell as shell;
#[doc(inline)]
pub use druid_shell::{kurbo, piet};

// the im crate provides immutable data structures that play well with druid
#[cfg(feature = "im")]
#[doc(inline)]
pub use im;

#[macro_use]
pub mod lens;

#[macro_use]
mod util;

mod app;
mod app_delegate;
mod bloom;
mod box_constraints;
mod command;
mod contexts;
mod core;
mod data;
mod dialog;
mod env;
mod event;
mod ext_event;
mod localization;
pub mod menu;
mod mouse;
pub mod scroll_component;
mod sub_window;
#[cfg(not(target_arch = "wasm32"))]
pub mod tests;
pub mod text;
pub mod theme;
pub mod widget;
mod win_handler;
mod window;

// Types from kurbo & piet that are required by public API.
pub use kurbo::{Affine, Insets, Point, Rect, Size, Vec2};
pub use piet::{Color, ImageBuf, LinearGradient, RadialGradient, RenderContext, UnitPoint};

// these are the types from shell that we expose; others we only use internally.
#[cfg(feature = "image")]
pub use shell::image;
pub use shell::keyboard_types;
pub use shell::{
    Application, Clipboard, ClipboardFormat, Code, Cursor, CursorDesc, Error as PlatformError,
    FileInfo, FileSpec, FormatId, HotKey, KbKey, KeyEvent, Location, Modifiers, Monitor,
    MouseButton, MouseButtons, RawMods, Region, Scalable, Scale, Screen, SysMods, TimerToken,
    WindowHandle, WindowLevel, WindowState,
};

pub use crate::core::WidgetPod;
pub use app::{AppLauncher, WindowConfig, WindowDesc, WindowSizePolicy};
pub use app_delegate::{AppDelegate, DelegateCtx};
pub use box_constraints::BoxConstraints;
pub use command::{sys as commands, Command, Notification, Selector, SingleUse, Target};
pub use contexts::{EventCtx, LayoutCtx, LifeCycleCtx, PaintCtx, UpdateCtx};
pub use data::Data;
pub use dialog::FileDialogOptions;
pub use env::{Env, Key, KeyOrValue, Value, ValueType, ValueTypeError};
pub use event::{Event, InternalEvent, InternalLifeCycle, LifeCycle};
pub use ext_event::{ExtEventError, ExtEventSink};
pub use lens::{Lens, LensExt};
pub use localization::LocalizedString;
pub use menu::{sys as platform_menus, Menu, MenuItem};
pub use mouse::MouseEvent;
pub use util::Handled;
pub use widget::{Widget, WidgetExt, WidgetId};
pub use win_handler::DruidHandler;
pub use window::{Window, WindowId};

#[cfg(not(target_arch = "wasm32"))]
pub(crate) use event::{StateCell, StateCheckFn};

#[deprecated(since = "0.8.0", note = "import from druid::text module instead")]
pub use piet::{FontFamily, FontStyle, FontWeight, TextAlignment};
#[deprecated(since = "0.8.0", note = "import from druid::text module instead")]
pub use text::{ArcStr, FontDescriptor, TextLayout};

/// The meaning (mapped value) of a keypress.
///
/// Note that in previous versions, the `KeyCode` field referred to the
/// physical position of the key, rather than the mapped value. In most
/// cases, applications should dispatch based on the value instead. This
/// alias is provided to make that transition easy, but in any case make
/// an explicit choice whether to use meaning or physical location and
/// use the appropriate type.
#[deprecated(since = "0.7.0", note = "Use KbKey instead")]
pub type KeyCode = KbKey;

#[deprecated(since = "0.7.0", note = "Use Modifiers instead")]
/// See [`Modifiers`](struct.Modifiers.html).
pub type KeyModifiers = Modifiers;
