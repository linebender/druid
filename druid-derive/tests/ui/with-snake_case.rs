#![allow(non_camel_case_types)]

use druid::*;

#[derive(Lens)]
struct my_thing {
    field: i32
}

fn main() {}
