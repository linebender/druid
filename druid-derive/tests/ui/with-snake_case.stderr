error: Lens implementations can only be derived from CamelCase types
 --> $DIR/with-snake_case.rs:6:8
  |
6 | struct my_thing {
  |        ^^^^^^^^
